libs/execution_context/ebpf_maps.c

```diff
@@ -616,9 +680,102 @@ _delete_object_hash_map(_In_ ebpf_core_map_t* map)
         ebpf_object_dereference_by_id(id, EBPF_OBJECT_MAP);
     }
 
+    if (object_map->inner_map_template != NULL) {
+        ebpf_object_release_reference(&object_map->inner_map_template->object);
+    }
+
     _delete_hash_map(map);
 }
 
+static ebpf_result_t
+_create_object_hash_map(
+    _In_ const ebpf_map_definition_in_memory_t* map_definition,
+    ebpf_handle_t inner_map_handle,
+    _Outptr_ ebpf_core_map_t** map)
+{
+    ebpf_core_map_t* local_map = NULL;
+    ebpf_result_t result = EBPF_SUCCESS;
+
+    EBPF_LOG_ENTRY();
+
+    *map = NULL;
+
+    result = _create_hash_map_internal(sizeof(ebpf_core_object_map_t), map_definition, NULL, &local_map);
+    if (result != EBPF_SUCCESS)
+        goto Exit;
+
+    ebpf_core_object_map_t* object_map = EBPF_FROM_FIELD(ebpf_core_object_map_t, core_map, local_map);
+    result = _associate_inner_map(object_map, inner_map_handle);
+    if (result != EBPF_SUCCESS)
+        goto Exit;
+
+    *map = local_map;
+    local_map = NULL;
+
+Exit:
+    if (result != EBPF_SUCCESS) {
+        _delete_object_hash_map(local_map);
+        local_map = NULL;
+    }
+
+    EBPF_RETURN_RESULT(result);
+}
+
+static ebpf_result_t
+_create_lru_hash_map(
+    _In_ const ebpf_map_definition_in_memory_t* map_definition,
+    ebpf_handle_t inner_map_handle,
+    _Outptr_ ebpf_core_map_t** map)
+{
+    ebpf_result_t retval = EBPF_SUCCESS;
+    ebpf_core_lru_map_t* lru_map = NULL;
+
+    UNREFERENCED_PARAMETER(inner_map_handle);
+
+    *map = NULL;
+
+    EBPF_LOG_ENTRY();
+
+    retval = _create_hash_map_internal(sizeof(ebpf_core_lru_map_t), map_definition, NULL, (ebpf_core_map_t**)&lru_map);
+    if (retval != EBPF_SUCCESS)
+        goto Exit;
+
+    // Note:
+    // ebpf_hash_table_t doesn't require synchronization as long as allocations
+    // are performed using the epoch allocator.
+    retval = ebpf_hash_table_create(
+        &lru_map->key_history,
+        ebpf_epoch_allocate,
+        ebpf_epoch_free,
+        lru_map->core_map.ebpf_map_definition.key_size,
+        sizeof(uint64_t),
+        lru_map->core_map.ebpf_map_definition.max_entries,
+        NULL);
+
+    if (retval != EBPF_SUCCESS)
+        goto Exit;
+
+    *map = &lru_map->core_map;
+
+Exit:
+    if (retval != EBPF_SUCCESS) {
+        ebpf_hash_table_destroy((ebpf_hash_table_t*)lru_map->core_map.data);
+        ebpf_epoch_free(lru_map);
+        lru_map = NULL;
+    }
+
+    EBPF_RETURN_RESULT(retval);
+}
+
+static void
+_delete_lru_hash_map(_In_ ebpf_core_map_t* map)
```

Comment:
Suggestion: I think it would be good to add the `_Post_invalid_` sal annotation to the delete APIs.  We do have it in ebpf_program_free_program_info() but not here yet.